### PR TITLE
Adds a -Werror debug Autotools test and updates build modes

### PR DESCRIPTION
**Pull request by derobins**
The Autotools build modes didn't look like they were doing the right thing, so I updated that while I was in there.

### CODE REVIEW

**Review by lrknox — APPROVED** [review]

**Comment by derobins** [comment]
There are some gcc 12 warnings that I was not seeing on my gcc 11 VM that cause failures with -Werror. I'll submit a separate PR with gcc 12 warning fixes and merge this PR after that one goes in.

File: .github/workflows/pr-check.yml [hunk 1]
```diff
@@ -81,6 +86,21 @@ jobs:
             parallel: disable
             toolchain: ""
             generator: "autogen"
+            flags: ""
+          # Build that ensures no new warnings sneak into the library.
+          - name: "-Werror Ubuntu Debug Autotools GCC"
+            artifact: "LinuxDBG.tar.xz"
+            os: ubuntu-latest
+            build_type: "debug"
```
Comment (byrnHDF) [inline .github/workflows/pr-check.yml:107]: Capital "D" - "Debug"

**Review by byrnHDF — APPROVED** [review]

File: .github/workflows/pr-check.yml (hunk 1 above)
Comment (derobins) [reply .github/workflows/pr-check.yml:107]: Not in the Autotools. We need --enable-build-mode=debug.

**Comment by derobins** [comment]
This PR should NOT get merged elsewhere. Only develop has been made warning-free.

**Comment by gnuoyd** [comment]
Re: commit [5b56fb3](https://github.com/HDFGroup/hdf5/pull/1833/commits/5b56fb362aca39b24095d38771ba9503b416cbc5), it looks like you may not be using `--enable-warnings-as-errors`. That will turn on the errors promotion without breaking `configure`.

File: configure.ac [hunk 2]
```diff
@@ -181,6 +183,14 @@ saved_user_JAVAFLAGS="$JAVAFLAGS"
 saved_user_LDFLAGS="$LDFLAGS"
 saved_user_CPPFLAGS="$CPPFLAGS"
 
+## Strip out -Werror from CFLAGS since that can cause checks to fail when
+## compiling the test program fails due to warnings
+CFLAGS="`echo $CFLAGS | sed -e 's/-Werror//g'`"
```
Comment (opoplawski) [inline configure.ac:188]: This breaks compilation on Fedora which uses `-Werror=format-security`

**Pull request merged**